Make Synapse use Cache auto tune

FSG-Cat opened the pull request:
This PR makes Synapse use cache Autotuning instead of relying on the low cache factor of 0.5 that is actually not recomended to be used by the devs them selfs(They recomend higher values as default and auto tune). Auto tune is more recomended because instead of just relying on the cache factor to control resource consumption we now have limits and targets allowing us to run a lot bigger cache sizes. 

Autotune requires jemalloc but that isnt a problem. The docker image we use already uses it so on that front we are ok. 

The current defaults could use some of the magic used in the Postgres autotune logic in the postgres role to adapt to system resources but these values that are in this PR will work for users that arent very ram constrained. 


Current defaults are a mix of the upstream default example and some bigger memory values that are supplied by people who have experience with this. 

This branch successfully deploys on test system and starts.

FSG-Cat committed:
Modify Synapse Cache Factor to use Auto Tune

Synapse has the ability to as it calls in its config auto tune caches.

This ability lets us set very high cache factors and then instead limit our resource use.

Defaults for this commit are 1/10th of what Element apparently runs for EMS stuff and matrix.org on Cache Factor and upstream documentation defaults for auto tune.

FSG-Cat committed:
Add vars to Synapse main.yml to control cache related config

This commit adds various cache related vars to main.yml for Synapse.

Some are auto tune and some are just adding explicit ways to control upstream vars.

FSG-Cat committed:
Updated Auto Tune figures

Autotuned figures have been bumped in consultation with other community members as to a reasonable level. Please note these defaults are more on the one of each workers side than they are on the monolith Side.

FSG-Cat committed:
Fix YML Error

The playbook is not happy with the previous state of this patch so this commit hopefully fixes it

FSG-Cat commented:
So im getting this error from trying to run this and i cant figure out where its comming from. So that atleast is something.
```
  msg: |-
    An unhandled exception occurred while templating '{{ matrix_synapse_configuration_yaml | from_yaml | combine(matrix_synapse_configuration_extension, recursive=True) }}'. Error was a <class 'yaml.parser.ParserError'>, original message: while parsing a block mapping
      in "<unicode string>", line 23, column 1
    did not find expected key
      in "<unicode string>", line 760, column 3
```

cvwright commented on roles/custom/matrix-synapse/defaults/main.yml:
```diff
@@ -395,9 +395,16 @@ matrix_synapse_event_cache_size: "100K"
+matrix_synapse_caches_expire_caches: true
+matrix_synapse_caches_cache_entry_ttl: 30m
+matrix_synapse_caches_sync_response_cache_duration: 2m
+matrix_synapse_caches_autotuning_max_cache_memory_usage: 2048M
+matrix_synapse_caches_autotuning_target_cache_memory_usage: 1024M
+matrix_synapse_caches_autotuning_min_cache_ttl: 5m
```
Maybe you want to make autotuning optional, like make a `matrix_synapse_caches_autotuning_enabled` and set it to `false` by default.

Then you wouldn't need to find values that work for everyone in every possible setup.  If a user is going to bother to enable autotune, then you can assume they will also put in the work to choose reasonable values.

FSG-Cat replied:
That is a resonable proposal to solve that concern. But we would still need to solve that this branch doesnt correctly deploy for unknown reasons. Like try to deploy this branch to a test system and it should complain.

spantaleev committed:
Add to_json to various Synapse tuning related configs

spantaleev committed:
Fix incorrect indication in homeserver.yaml.j2

spantaleev committed:
Minor cleanups

FSG-Cat committed:
Synapse Cache Autotuning Documentation

FSG-Cat commented on roles/custom/matrix-synapse/defaults/main.yml:
```diff
@@ -546,8 +546,16 @@ matrix_synapse_event_cache_size: "100K"
+matrix_synapse_caches_expire_caches: true
+matrix_synapse_caches_cache_entry_ttl: "30m"
+matrix_synapse_caches_sync_response_cache_duration: "2m"
+matrix_synapse_caches_autotuning_max_cache_memory_usage: "2048M"
```
Oh yes this is a known problem with this PR and its addressed in the PR description maby not to well and has been mentioned in the matrix room when discussing this PR that this is indeed a good idea.

I will look at taking a stab at implementing this in the role as to fix this flaw. So Thanks for this comment.

FSG-Cat replied:
This should be resolved in the latest commits as we now use 1/8 and 1/16 as targets up to the previous values.

FSG-Cat replied:
After a conversation with @tcpipuk about the topic the conclusion has been reached that you dont rly want these numbers to go higher than X. Especially as these numbers are actually allowed to be reached per worker. And also the fact that its useless to allow this to scale to infinity as the cache factor creates its own max and autotune is essentially GC for that. 

And also that in the end all these values are sensible defaults that are intended to be good enough for those who dont touch these variables they are not intended to actually be a good idea to run for all users. Users who know what they are doing should be pushing way more advanced cache control settings than this PR has. Doing stuff like tuning the sizes of the specific types of caches etc and even preferably doing these settings per worker type or per worker.

tcpipuk replied:
I personally would increase the "matrix_synapse_caches_cache_entry_ttl" because the 30m default is there for servers that aren't autotuning.

In an ideal world jemalloc would clear out anything that's not regularly used, so any objects that reach a long TTL would be accessed frequently, and it can really improve response times to let those live longer (especially on workers that do federation) but I guess it depends how far you want to deviate from the defaults?

I guess it could just be covered by a couple of lines in the docs to explain that setting 4-8h can massively improve response times on quieter servers, as the workers don't end up dumping everything when the user closes their client for an hour 🙂

FSG-Cat committed:
Upgrade Synapse Cache Autotune to auto configure memory use

FSG-Cat committed:
Update Synapse Tuning docs to reflect automatic memory use configuration

FSG-Cat committed:
Fix Linting errors in synapses main.yml

spantaleev committed:
Rename variables for consistency (matrix_synapse_caches_autotuning_* -> matrix_synapse_cache_autotuning_*)

spantaleev committed:
Remove FIX ME comment about Synapse's `cache_autotuning`

`docs/maintenance-synapse.md` and `roles/custom/matrix-synapse/defaults/main.yml`
already contains documentation about these variables and the default values we set.

spantaleev committed:
Improve "Tuning caches and cache autotuning" documentation for Synapse

spantaleev committed:
Announce larger Synapse caches and cache auto-tuning

spantaleev committed:
Merge branch 'master' into Synapse-Caches-auto-tune
